File: core/session_backup.py
# core/session_backup.py
"""크래시 복구용 세션 백업(cache/session/session_backup.json) — 읽기·쓰기·정상 종료 표시.

Vue 가 편집 2.5초 뒤·30초마다 {tab, prompt, negative} 를 저장한다. 예전엔 복구 제안이
'현재 메인 프롬프트가 비어 있을 때'만 떴는데, 부팅의 load_settings 가 prompt_settings.json 으로
프롬프트를 늘 채우므로 사실상 뜨지 않았고, 곧이어 주기 저장이 백업을 낡은 프롬프트로 덮어
크래시 직전 편집이 영구히 사라졌다(감사 #160).

이제 정상 종료(_save_shutdown_state)가 백업에 ``clean: true`` 를 표시한다. 다음 부팅에서
``clean`` 이 아니고 백업 프롬프트가 현재 값과 다르면 크래시로 남은 작업이다
(frontend/src/utils/sessionBackup.ts shouldOfferSessionRestore). mtime 비교는 쓰지 않는다 —
부팅 중 연결 경로의 save_settings 가 prompt_settings.json 의 mtime 을 갱신한다.
"""
from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Mapping, Optional

SESSION_BACKUP_NAME = "session/session_backup.json"
_LEGACY_PATH = "config/session_backup.json"


def session_backup_path() -> Path:
    from core.storage_paths import cache_file

    return cache_file(SESSION_BACKUP_NAME, legacy_paths=_LEGACY_PATH)


def read_session_backup(path: Optional[Path] = None) -> dict:
    """백업 dict — 없거나 깨졌으면 ``{}``."""
    target = Path(path) if path is not None else session_backup_path()
    try:
        if not target.exists():
            return {}
        data = json.loads(target.read_text(encoding="utf-8") or "{}")
    except (OSError, ValueError):
        return {}
    return data if isinstance(data, dict) else {}


def write_session_backup(payload: Mapping[str, Any], path: Optional[Path] = None) -> dict:
    """편집 중 백업 — 늘 ``clean: false`` 로 쓴다(정상 종료만 clean 을 세운다)."""
    if not isinstance(payload, Mapping):
        raise ValueError("세션 payload는 JSON 객체여야 합니다")
    data = dict(payload)
    data["clean"] = False
    from utils.atomic_json import atomic_write_json

    target = Path(path) if path is not None else session_backup_path()
    atomic_write_json(str(target), data, indent=None)
    return data


def mark_session_clean(path: Optional[Path] = None) -> bool:
    """정상 종료 표시 — 다음 부팅이 이 백업을 복구 대상으로 보지 않는다. 백업이 없으면 no-op."""
    target = Path(path) if path is not None else session_backup_path()
    data = read_session_backup(target)
    if not data or data.get("clean") is True:
        return False
    data["clean"] = True
    from utils.atomic_json import atomic_write_json

    atomic_write_json(str(target), data, indent=None)
    return True

File: frontend/src/utils/sessionBackup.ts
/**
 * 크래시 복구 백업(cache/session/session_backup.json)의 순수 규칙 (감사 #160).
 *
 * 예전 조건 '현재 메인 프롬프트가 비었을 때만 제안'은 부팅의 load_settings 가 prompt_settings.json
 * 으로 프롬프트를 늘 채우므로 사실상 성립하지 않았고, 곧이어 2.5초/30초 저장이 백업을 낡은 값으로
 * 덮어 크래시 직전 편집을 영구히 지웠다. 이제
 * - 정상 종료는 백업에 ``clean: true`` 를 남긴다(Python core/session_backup.mark_session_clean).
 * - ``clean`` 이 아니고, 백업 프롬프트가 있으며, 지금 화면 값(메인·네거티브)과 다르면 제안한다.
 * - 사용자가 고르기 전엔 백업을 덮지 않는다(composables/useSessionRestore.ts).
 * mtime 은 쓰지 않는다 — 부팅 중 연결 경로의 save_settings 가 prompt_settings.json 을 새로 쓴다.
 */

export interface SessionBackup {
  tab?: string
  prompt?: string
  negative?: string
  /** 정상 종료 표시 — true 면 복구할 것이 없다 */
  clean?: boolean
  /** 저장 시각(ms) — 표시·진단용 */
  savedAt?: number
  [key: string]: unknown
}

export interface SessionPrompts {
  prompt?: string | null
  negative?: string | null
}

const text = (value: unknown): string => (typeof value === 'string' ? value.trim() : '')

/** JSON 응답 → 백업 dict. 깨졌거나 오류 응답({error})이면 null. */
export function parseSessionBackup(json: unknown): SessionBackup | null {
  let data: unknown = json
  if (typeof json === 'string') {
    try { data = JSON.parse(json || '{}') } catch { return null }
  }
  if (!data || typeof data !== 'object' || Array.isArray(data)) return null
  if ('error' in (data as Record<string, unknown>)) return null
  return data as SessionBackup
}

/** 복구를 제안할까 — 정상 종료가 아니고, 백업 프롬프트가 있고, 지금 화면과 다를 때. */
export function shouldOfferSessionRestore(
  backup: SessionBackup | null | undefined,
  current: SessionPrompts,
): boolean {
  if (!backup || backup.clean === true) return false
  const savedPrompt = text(backup.prompt)
  if (!savedPrompt) return false
  return savedPrompt !== text(current.prompt) || text(backup.negative) !== text(current.negative)
}

/** 저장할 백업 한 벌 — clean 은 Python 이 늘 false 로 쓴다. */
export function buildSessionBackup(
  tab: unknown,
  prompts: SessionPrompts,
  now: number = Date.now(),
): SessionBackup {
  return {
    tab: typeof tab === 'string' && tab ? tab : 't2i',
    prompt: typeof prompts.prompt === 'string' ? prompts.prompt : '',
    negative: typeof prompts.negative === 'string' ? prompts.negative : '',
    savedAt: now,
  }
}
